File: chelseashin/Programmers/월간코드챌린지시즌2/괄호회전하기.py
# 8분 소요
def solution(s):
    answer = 0
    for _ in range(len(s)):
        s = s[-1] + s[:-1]
        stack = []
        for x in s:
            if not stack:
                stack.append(x)
                continue
            if x == "]" and stack[-1] == "[":
                stack.pop()
            elif x == "}" and stack[-1] == "{":
                stack.pop()
            elif x == ")" and stack[-1] == "(":
                stack.pop()
            else:
                stack.append(x)
        if not stack:
            answer += 1
    return answer

print(solution("[](){}"))

File: hyeanwooRoh/programmers/codeChallenge/Season2/괄호회전하기.java
// 00:02 ~ 00:40

import java.util.*;

public class Main {
	public static void main(String[] args) {
    System.out.println(solution("[](){}"));
    System.out.println(solution("}]()[{"));
    System.out.println(solution("[)(]"));
    System.out.println(solution("}}}"));
	}

  public static int solution(String s) {
    int answer = 0;
    for (int i = 0; i < s.length(); i++) {
      if(stacking(s, i)) answer++;
    }

    return answer;
  }

  public static boolean stacking(String s, int start) {
    int cycleCnt = 0;
    int idx = start;

    Stack<Character> stack = new Stack<>();
    while (cycleCnt < s.length()) {
      char ch = s.charAt(idx);
      if (ch == '[' || ch == '{' || ch == '(') {
        stack.push(ch);
      } else {
        if (stack.empty() || !correctBraket(stack.pop(), ch)) {
          return false;
        }
      }
      
      idx = idx + 1 == s.length() ? 0 : idx + 1;
      cycleCnt++;
    }

    return stack.empty();
  }

  public static boolean correctBraket(char ch1, char ch2) {
    switch(ch1) {
      case '[':
        return ch2 == ']';
      case '{':
        return ch2 == '}';
      case '(':
        return ch2 == ')';
    }

    return false;
  }
}
